pkg/util/mount/mount_linux.go

```diff
@@ -425,3 +434,129 @@ func (mounter *SafeFormatAndMount) getDiskFormat(disk string) (string, error) {
 	// and MD RAID are reported as FSTYPE and caught above).
 	return "unknown data, probably partitions", nil
 }
+
+// isShared returns true, if given path is on a mount point that has shared
+// mount propagation.
+func isShared(path string, filename string) (bool, error) {
+	infos, err := getMountInfo(filename)
+	if err != nil {
+		return false, err
+	}
+
+	// process /proc/xxx/mountinfo in backward order and find the first mount
+	// point that is prefix of 'path' - that's the mount where path resides
+	var info *mountInfo
+	for i := len(infos) - 1; i >= 0; i-- {
+		if strings.HasPrefix(path, infos[i].mountPoint) {
+			info = &infos[i]
+			break
+		}
+	}
+	if info == nil {
+		return false, fmt.Errorf("cannot find mount point for %q", path)
+	}
+
+	// parse optional parameters
+	for _, opt := range info.optional {
+		if strings.HasPrefix(opt, "shared:") {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+type mountInfo struct {
+	root, mountPoint string
+	// list of "optional parameters", mount propagation is one of them
+	optional []string
+}
+
+// getMountInfo reads /proc/xxx/mountinfo and makes sure it did not change
+// between reads. This protects us from kernel adding/removing lines there
+// between individual read() syscalls.
+func getMountInfo(filename string) ([]mountInfo, error) {
+	// Read the file until we get the same content twice
+	oldInfo, err := parseMountInfo(filename)
+	if err != nil {
+		return []mountInfo{}, err
+	}
+	for {
```

Comment:
I wish we could make this a generic utility for reading many kernel APIs that return data more than a page in size.